### PR TITLE
Upgrade Resteasy v4

### LINKED ISSUE #10916: Upgrade Resteasy v4
### Description

Both legacy and main distribution are using Resteasy v4 now.

However, we are still compiling using Resteasy v3.

We should now be able to fully upgrade to Resteasy v4.

### Discussion

https://github.com/keycloak/keycloak/issues/10907

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #10916

* Change version in parent pom to the version used in Wildfly
* Change poms to use `resteasy-core` instead of `resteasy-jaxrs`
* Wildfly modules do not need to change because Wildfly is still able to resolve `org.jboss.resteasy:resteasy-jaxrs` module dependency, which points to `resteasy-core` module.
* Updates to the list of providers defined for the admin client.

### CODE REVIEW

File: testsuite/model/pom.xml
Comment (pedroigor) [inline testsuite/model/pom.xml:40]: The version is the same as the transitive dependency in `resteasy-jaxrs`. But not sure if we actually need or want to have the properties in the parent pom.

Comment (hmlnarik) [reply testsuite/model/pom.xml:40]: Is it used in model tests at all? Is it necessary to be added?

Comment (pedroigor) [reply testsuite/model/pom.xml:40]: That is what I asking but looks like you don't know either. Thanks.

Comment (hmlnarik) [reply testsuite/model/pom.xml:40]: I got response offline:

> Yeah, there is a test using an annotation
> @Guard or something
> org.infinispan.persistence.manager.PersistenceManagerImpl

So I'm fine for this PR. See #10924 for revert requirement

Comment (pedroigor) [reply testsuite/model/pom.xml:40]: Tks

**Comment by pedroigor** [comment]
@abstractj Looks like test suite is passing now. The last failure was related to the providers defined in the `javax.ws.rs.ext.Providers` in the `keycloak-admin-cli` module. This module does not use JAX-RS at all and I removed the file from there.

In v4, enabling the GZIP providers was causing responses to be encoded twice. For instance, serving `keycloak.js`.

**Comment by stianst** [comment]
This needs to wait to after Keycloak 18, as EAP 7.4 is using RestEasy 3

File: integration/client-cli/admin-cli/src/main/resources/META-INF/services/javax.ws.rs.ext.Providers
Comment (stianst) [inline integration/client-cli/admin-cli/src/main/resources/META-INF/services/javax.ws.rs.ext.Providers:1]: Why is this no longer needed?

Comment (stianst) [inline integration/client-cli/admin-cli/src/main/resources/META-INF/services/javax.ws.rs.ext.Providers:1]: Why is this no longer needed?

**Review by stianst — COMMENTED** [review]
Looks okay to me, a few question/comments though:

* We need to run internal pipelines against this
* Can this cause some breaking changes to adapters/clients? Do we still support RestEasy 3 at all on the client side, for example in the admin client? Is upgrade to HttpClient a potentially breaking change? I presume we'll at the very least will require some notice around this in migration guide?
* Unsure about the SPNEGO support in HttpClientBuilder. Do we use this for Kerberos support? If so we need to make sure this stuff is tested
* Is this really two upgrades in one? HttpClient upgrade as well as RestEasy? I presume that's because RestEasy 4 requires a newer HttpClient?

File: integration/client-cli/admin-cli/src/main/resources/META-INF/services/javax.ws.rs.ext.Providers
Comment (pedroigor) [reply integration/client-cli/admin-cli/src/main/resources/META-INF/services/javax.ws.rs.ext.Providers:1]: The `keycloak-admin-cli` does not use JAX-RS. See comment https://github.com/keycloak/keycloak/pull/10917#issuecomment-1081904455.

**Comment by pedroigor** [comment]
> Looks okay to me, a few question/comments though:
> 
> * We need to run internal pipelines against this
> * Can this cause some breaking changes to adapters/clients? Do we still support RestEasy 3 at all on the client side, for example in the admin client? Is upgrade to HttpClient a potentially breaking change? I presume we'll at the very least will require some notice around this in migration guide?

I don't think it will be a breaking change. The reason is that adapters and clients are wrapping the dependency and we don't really expected people relying on internals. 

But yeah, it deserves a release note at the very minimum.

> * Unsure about the SPNEGO support in HttpClientBuilder. Do we use this for Kerberos support? If so we need to make sure this stuff is tested

AFAIK, the testsuite (GHA) also checks SPNEGO. I can double check.

> * Is this really two upgrades in one? HttpClient upgrade as well as RestEasy? I presume that's because RestEasy 4 requires a newer HttpClient?

`apache:http-client` should be the same: `4.5.13`.

**Comment by pedroigor** [comment]
@stianst Can we move forward with this one?

**Review by stianst — DISMISSED** [review]

**Comment by stianst** [comment]
@pedroigor LGTM, would be good to have updates to release notes added before we merge it, and also would be good to get @mposolda or someone else to review this one as well

**Comment by mabartos** [comment]
@abstractj Yes please, the pipeline should be executed. I'll do it and share the results ASAP.

**Comment by mabartos** [comment]
Besides that, the `HttpClientBuilder` is duplicated also for the SAML adapters and AFAIK, there was some convention that when you change sth in one of those classes, you should provide those changes even to the other one (if it's not specific to the particular protocol). As we don't have any better solution for that right now, we should keep the consistency in those classes at least by the mentioned approach. 

As those adapters will be present, at least, for some time and the `HttpClientBuilder` is quite important in the context of those adapters, it'd be nice to provide those changes there as well.

@pedroigor @stianst WDYT?

**Comment by mabartos** [comment]
Pipelines results seem quite good, but there are some failures for adapter tests and Tomcat. I'll verify it.

Adapter pipeline: https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/1056/
Server pipeline: https://keycloak-jenkins.com/job/universal-test-pipeline-server/1637/

File: adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java
Comment (ahus1) [inline adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java:573]: I see that the trustmanager is disabled here. I'm lacking a comment why this would be a good idea here, and also doing this as a default without an option to re-configure it. I know this hasn't been changed in this PR, so it might a new issue. 

I had a look into [keycloak-documentation](https://github.com/keycloak/keycloak-documentation/blob/main/securing_apps/topics/oidc/java/installed-adapter.adoc) to find out what this class is trying to achieve, and didn't find a reason why not to check the certificates of the Keycloak server.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Some comments about timeouts and trust manager. While the timeouts could be handled in this issue, the trustmanager comment should probably be handled outside of the issue. 

Even both could be handled in a separate PR if we choose to merge this PR now.

**Review by mabartos — CHANGES_REQUESTED** [review]
@pedroigor Unfortunately, those failures for Tomcat are caused by this PR.

Log: https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/1056/execution/node/2350/log/?consoleFull

Steps to reproduce:
1) build Tomcat app server (`servers/app-server/tomcat -Papp-server-tomcat9`)
2) execute in the base module 
```
mvn clean install -Djs.browser=chrome -Dtest=DemoServletsAdapterTest -Pauth-server-wildfly -Papp-server-tomcat9
```

```
2022-06-08 20:04:32  08-Jun-2022 14:04:30.673 INFO [main] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/home/jenkins/workspace/universal-test-pipeline-adapters/adapter_main_tests-tomcat9-tls/keycloak/testsuite/integration-arquillian/tests/base/target/containers/app-server-tomcat9/webapps/ROOT]
2022-06-08 20:04:32  08-Jun-2022 14:04:30.922 SEVERE [main] org.apache.catalina.startup.HostConfig.deployDirectory Error deploying web application directory [/home/jenkins/workspace/universal-test-pipeline-adapters/adapter_main_tests-tomcat9-tls/keycloak/testsuite/integration-arquillian/tests/base/target/containers/app-server-tomcat9/webapps/ROOT]
2022-06-08 20:04:32  	java.lang.IllegalStateException: Error starting child
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:720)
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:690)
2022-06-08 20:04:32  		at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:705)
2022-06-08 20:04:32  		at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1133)
2022-06-08 20:04:32  		at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1867)
2022-06-08 20:04:32  		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
2022-06-08 20:04:32  		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2022-06-08 20:04:32  		at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
2022-06-08 20:04:32  		at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:112)
2022-06-08 20:04:32  		at org.apache.catalina.startup.HostConfig.deployDirectories(HostConfig.java:1045)
2022-06-08 20:04:32  		at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:429)
2022-06-08 20:04:32  		at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1576)
2022-06-08 20:04:32  		at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:309)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:423)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:366)
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:936)
2022-06-08 20:04:32  		at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:841)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1384)
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1374)
2022-06-08 20:04:32  		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2022-06-08 20:04:32  		at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
2022-06-08 20:04:32  		at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134)
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:909)
2022-06-08 20:04:32  		at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:262)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
2022-06-08 20:04:32  		at org.apache.catalina.core.StandardService.startInternal(StandardService.java:421)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
2022-06-08 20:04:32  		at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:930)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
2022-06-08 20:04:32  		at org.apache.catalina.startup.Catalina.start(Catalina.java:633)
2022-06-08 20:04:32  		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-06-08 20:04:32  		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-06-08 20:04:32  		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-06-08 20:04:32  		at java.lang.reflect.Method.invoke(Method.java:498)
2022-06-08 20:04:32  		at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:343)
2022-06-08 20:04:32  		at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:474)
2022-06-08 20:04:32  	Caused by: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[]]
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.handleSubClassException(LifecycleBase.java:440)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:198)
2022-06-08 20:04:32  		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:717)
2022-06-08 20:04:32  		... 37 more
2022-06-08 20:04:32  	Caused by: java.lang.NoClassDefFoundError: org/jboss/resteasy/spi/NotImplementedYetException
2022-06-08 20:04:32  		at java.lang.Class.forName0(Native Method)
2022-06-08 20:04:32  		at java.lang.Class.forName(Class.java:348)
2022-06-08 20:04:32  		at org.apache.catalina.startup.WebappServiceLoader.loadServices(WebappServiceLoader.java:188)
2022-06-08 20:04:32  		at org.apache.catalina.startup.WebappServiceLoader.load(WebappServiceLoader.java:159)
2022-06-08 20:04:32  		at org.apache.catalina.startup.ContextConfig.processServletContainerInitializers(ContextConfig.java:1669)
2022-06-08 20:04:32  		at org.apache.catalina.startup.ContextConfig.webConfig(ContextConfig.java:1137)
2022-06-08 20:04:32  		at org.apache.catalina.startup.ContextConfig.configureStart(ContextConfig.java:774)
2022-06-08 20:04:32  		at org.apache.catalina.startup.ContextConfig.lifecycleEvent(ContextConfig.java:301)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
2022-06-08 20:04:32  		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5051)
2022-06-08 20:04:32  		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
2022-06-08 20:04:32  		... 38 more
2022-06-08 20:04:32  	Caused by: java.lang.ClassNotFoundException: org.jboss.resteasy.spi.NotImplementedYetException
2022-06-08 20:04:32  		at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
2022-06-08 20:04:32  		at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
2022-06-08 20:04:32  		at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
2022-06-08 20:04:32  		... 49 more
```

File: adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java
Comment (pedroigor) [reply adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java:573]: @ahus1 Thanks, updated.

File: adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java
Comment (pedroigor) [reply adapters/oidc/installed/src/main/java/org/keycloak/adapters/installed/KeycloakInstalled.java:575]: Yeah, I would deal with it in a separate issue. But I wouldn't worry about that until we have any issue reported by the community.

This adapter in particular is mainly developed by community and limited to very specific use cases.

**Comment by pedroigor** [comment]
@mabartos @ahus1 Thanks for your time on this one.

@mabartos I'll look at the error you mentioned. Indeed it seems related but no clue what can be by looking at the stacktrace.

**Comment by pedroigor** [comment]
@mabartos I was able to reproduce and fix the issue by removing this dependency from the tomcat classpath:

```
<artifactItem>
    <groupId>org.jboss.resteasy</groupId>
    <artifactId>resteasy-servlet-initializer</artifactId>
    <version>${tomcat.resteasy.version}</version>
</artifactItem>
```

Would you be able to check it again?

**Review by mabartos — COMMENTED** [review]
@pedroigor Thanks for fixing the issue with Tomcat. 

I tried to reproduce it and those tests passed for Tomcat 8 and Tomcat 9. I don't think is necessary to execute pipeline tests, as only the Tomcat module was affected. Those changes were locally verified and the correctness is confirmed.

Regarding the mentioned changes for SAML `HttpClientBuilder`, I'd like to resolve it in some follow-up task in order to speed up the process as it's not a blocking issue.
Could you please create an issue for it?

**Comment by mabartos** [comment]
@pedroigor What's the current status? Anything else need to be done?

**Review by abstractj — DISMISSED** [review]

**Review by mabartos — DISMISSED** [review]
Haven't got a chance to take a look at the code more deeply, but from a QE perspective, I can approve this PR. Pipeline tests passed and the necessary manual testing of a particular component passed as well.

@pedroigor Please, don't forget to create a follow-up issue to apply the `HttpClientBuilder` changes also to the representation of `HttpClientBuilder` in the `adapters/saml` module. Thanks!

It'd be also good to retry the GH actions for `Base testsuite (undertow-map, group1)` combination.

**Review by ahus1 — DISMISSED** [review]
My previous comments have been addressed.

**Comment by abstractj** [comment]
@pedroigor could you please double-check the failures on the Keycloak CI / Base testsuite? Not sure if they are related with those changes.

**Comment by pedroigor** [comment]
@abstractj Does not seem related. Running the failing tests.

**Comment by ahus1** [comment]
The current run looks a lot better, all undertow tests completed. The quickstart test seems unrelated to this change and flaky. Maybe a maintainer could rerun it.

As I was involved in preparing this PR on the last yards (fixing a double dependency in the test suite), I am hesitant to approve it. Still, I approve the rest of it.

**Comment by pedroigor** [comment]
@ahus1 Thanks for the fix. PR rebased and squashed.

**Review by ahus1 — APPROVED** [review]
Approving Pedros' parts. Thanks for the PR, happy to join this teamwork!

**Comment by mabartos** [comment]
Started a pipeline run to be sure.

https://keycloak-build-pipeline.com/job/keycloak-build-pipeline/808/

**Review by abstractj — APPROVED** [review]

**Comment by abstractj** [comment]
Fantastic work! I'm waiting for @mabartos feedback, so we can merge.

**Review by mabartos — APPROVED** [review]
LGTM! Failures in the server and adapter pipeline are not related to this PR.

**Pull request merged**
